### ISSUE TITLE
Key cache internal load throws a NullPointerException

### ISSUE THREAD

**Issue by johnmaia** [FusionAuth/fusionauth-issues#326]
### Description 
After a few hours running `fusionauth` (`v1.10.1`) flawlessly,  the application reached the system's memory limit and throws a `NullPointerException` from  the `KeyCacheLoader` `internalLoad` method.

This happened on our test environment where a single instance of `fusionauth` deals with over 4k tenants, 4k applications and over 14k users. We've been progressively increasing the system's memory and this happened when gave it 4GB of RAM.

```
FUSIONAUTH_MEMORY=4096M
```

We suspect that `fusionauth` stores too much data in memory as cache every time a new tenant is created. If this is true, `fusionauth` isn't scalable and it won't properly work in a use case that deals with large amount of `tenants`.

We've been dealing with `fusionauth` memory issues for a long time now. On version `1.7.2` the error was diferent, but it happened in the under same conditions.

### Steps to reproduce
1. Run `fusionauth` `v1.10.1`
2. Reach the System's memory limit
3. The following `NullPointerException` is thrown:

```
---------------------------------------------------------------------------------------------------------
--------------------------------- Starting FusionAuth version [1.10.1] ----------------------------------
---------------------------------------------------------------------------------------------------------

Oct 09, 2019 5:36:27.121 PM INFO  io.fusionauth.api.plugin.guice.PluginModule - No plugins found
Oct 09, 2019 5:36:27.363 PM INFO  io.fusionauth.api.service.system.NodeService - Starting Master node with Id [cc8717f5-5e25-47c3-9c2c-e99d8000bfe5]
Oct 09, 2019 5:36:28.135 PM INFO  io.fusionauth.api.configuration.PropertiesFileFusionAuthConfiguration - Loading FusionAuth configuration file [/usr/local/fusionauth/fusionauth-app/apache-tomcat/../../config/fusionauth.properties]
Oct 09, 2019 5:36:28.135 PM INFO  io.fusionauth.api.configuration.PropertiesFileFusionAuthConfiguration - Dynamically set property [fusionauth-app.public-url] set to [http://localhost:9011]
Oct 09, 2019 5:36:28.135 PM INFO  io.fusionauth.api.configuration.PropertiesFileFusionAuthConfiguration - FusionAuth Runtime Mode [Production]
Oct 09, 2019 5:36:28.138 PM INFO  com.inversoft.jdbc.hikari.DataSourceProvider - Connecting to PostgreSQL database at [jdbc:postgresql://database-main:5432/fusionauth]
Oct 09, 2019 5:36:28.150 PM INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Starting...
Oct 09, 2019 5:36:28.217 PM INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Start completed.
Oct 09, 2019 5:36:29.459 PM INFO  com.inversoft.scheduler.DefaultScheduler - Starting up scheduler
Oct 09, 2019 5:36:29.460 PM INFO  com.inversoft.scheduler.DefaultScheduler - Scheduler is running
Oct 09, 2019 5:36:44.562 PM INFO  io.fusionauth.api.service.system.NodeService - Add node with address [http://localhost:9011]
Oct 09, 2019 5:36:44.850 PM INFO  com.inversoft.migration.Migrator - Run Cleanup [io.fusionauth.api.migration.guice.Migration_1_6_0]
Oct 09, 2019 5:36:44.852 PM INFO  com.inversoft.migration.Migrator - Run Cleanup [io.fusionauth.api.migration.guice.Migration_1_7_0]
Oct 09, 2019 5:36:44.855 PM INFO  com.inversoft.migration.Migrator - Run Cleanup [io.fusionauth.api.migration.guice.Migration_1_8_0]
Oct 09, 2019 5:36:44.857 PM INFO  com.inversoft.migration.Migrator - Run Cleanup [io.fusionauth.api.migration.guice.Migration_1_9_2]
Oct 09, 2019 5:36:44.859 PM INFO  com.inversoft.migration.Migrator - Run Cleanup [io.fusionauth.api.migration.guice.Migration_1_10_0]
09-Oct-2019 17:36:44.906 INFO [main] org.apache.coyote.AbstractProtocol.start Starting ProtocolHandler ["http-nio-9011"]
09-Oct-2019 17:36:44.923 INFO [main] org.apache.coyote.AbstractProtocol.start Starting ProtocolHandler ["https-jsse-nio-9013"]
09-Oct-2019 17:36:44.927 INFO [main] org.apache.coyote.AbstractProtocol.start Starting ProtocolHandler ["ajp-nio-9019"]
09-Oct-2019 17:36:44.929 INFO [main] org.apache.catalina.startup.Catalina.start Server startup in 24698 ms
Oct 09, 2019 5:37:20.692 PM INFO  io.fusionauth.api.service.search.ElasticSearchClientProvider - Connecting to FusionAuth Search Engine at [http://elk-elasticsearch-client.elk.svc.cluster.local:9200]
Oct 09, 2019 5:37:58.845 PM INFO  io.fusionauth.api.service.system.NodeService - Remove node [http://localhost:9011] because it has not checked in for the last 107 seconds.
Oct 09, 2019 5:42:30.840 PM INFO  io.fusionauth.api.service.system.NodeService - Became the master Node with Id [cc8717f5-5e25-47c3-9c2c-e99d8000bfe5] at timestamp [2019-10-09T17:42:30.840Z]. Took over from [77ed4342-50eb-461a-8b89-f1afa439cc2a] who last updated at [2019-10-09T17:34:13.613Z]
Oct 10, 2019 10:50:46.729 AM ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown
java.lang.NullPointerException: null
	at io.fusionauth.api.service.cache.KeyCacheLoader.internalLoad(KeyCacheLoader.java:73)
	at com.inversoft.cache.BaseCacheLoader.load(BaseCacheLoader.java:21)
	at io.fusionauth.app.action.api.cache.ReloadAction.post(ReloadAction.java:40)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.primeframework.mvc.util.ReflectionUtils.invoke(ReflectionUtils.java:431)
	at org.primeframework.mvc.action.DefaultActionInvocationWorkflow.execute(DefaultActionInvocationWorkflow.java:84)
	at org.primeframework.mvc.action.DefaultActionInvocationWorkflow.perform(DefaultActionInvocationWorkflow.java:64)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.validation.DefaultValidationWorkflow.perform(DefaultValidationWorkflow.java:47)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.security.DefaultSecurityWorkflow.perform(DefaultSecurityWorkflow.java:81)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.parameter.DefaultPostParameterWorkflow.perform(DefaultPostParameterWorkflow.java:50)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.content.DefaultContentWorkflow.perform(DefaultContentWorkflow.java:52)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.parameter.DefaultParameterWorkflow.perform(DefaultParameterWorkflow.java:57)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.parameter.DefaultURIParameterWorkflow.perform(DefaultURIParameterWorkflow.java:102)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.scope.DefaultScopeRetrievalWorkflow.perform(DefaultScopeRetrievalWorkflow.java:58)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.message.DefaultMessageWorkflow.perform(DefaultMessageWorkflow.java:45)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at io.fusionauth.app.primeframework.FrontEndTenantWorkflow.perform(FrontEndTenantWorkflow.java:66)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.action.DefaultActionMappingWorkflow.perform(DefaultActionMappingWorkflow.java:126)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.workflow.StaticResourceWorkflow.perform(StaticResourceWorkflow.java:97)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.parameter.RequestBodyWorkflow.perform(RequestBodyWorkflow.java:89)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.security.DefaultSavedRequestWorkflow.perform(DefaultSavedRequestWorkflow.java:57)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at io.fusionauth.app.primeframework.CORSFilter.handleNonCORS(CORSFilter.java:289)
	at io.fusionauth.app.primeframework.CORSFilter.doFilter(CORSFilter.java:250)
	at io.fusionauth.app.primeframework.CORSRequestWorkflow.perform(CORSRequestWorkflow.java:48)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at io.fusionauth.app.primeframework.FusionAuthMVCWorkflow.perform(FusionAuthMVCWorkflow.java:88)
	at org.primeframework.mvc.workflow.DefaultWorkflowChain.continueWorkflow(DefaultWorkflowChain.java:44)
	at org.primeframework.mvc.servlet.FilterWorkflowChain.continueWorkflow(FilterWorkflowChain.java:50)
	at org.primeframework.mvc.servlet.PrimeFilter.doFilter(PrimeFilter.java:84)
	at com.inversoft.maintenance.servlet.MaintenanceModePrimeFilter.doFilter(MaintenanceModePrimeFilter.java:59)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at com.inversoft.servlet.UTF8Filter.doFilter(UTF8Filter.java:27)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:199)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:493)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:137)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:798)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:808)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1498)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Oct 10, 2019 10:50:46.746 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:50:46.834 AM ERROR com.inversoft.scheduler.LogAndRetainFailureHandler - The scheduled service [class io.fusionauth.api.service.cache.KeyCacheLoader] failed but will be re-run.
java.lang.NullPointerException: null
	at io.fusionauth.api.service.cache.KeyCacheLoader.internalLoad(KeyCacheLoader.java:73)
	at com.inversoft.cache.BaseCacheLoader.load(BaseCacheLoader.java:21)
	at io.fusionauth.api.service.cache.KeyCacheLoader.run(KeyCacheLoader.java:55)
	at com.inversoft.scheduler.InjectableRunnable.run(InjectableRunnable.java:37)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Oct 10, 2019 10:50:47.194 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:50:47.529 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:51:09.468 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:51:51.707 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:51:56.716 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:51:58.948 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:01.261 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:01.345 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:05.510 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:05.666 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:05.698 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:08.997 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:09.366 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:09.564 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:13.511 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:13.649 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:13.694 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:13.960 AM ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown
java.lang.NullPointerException: null
Oct 10, 2019 10:52:17.395 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:17.602 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
Oct 10, 2019 10:52:17.610 AM ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown
java.lang.NullPointerException: null
Oct 10, 2019 10:52:23.543 AM ERROR io.fusionauth.api.service.cache.DistributedCacheNotifier - Failed to request a cache reload for [http://localhost:9011]. Exception message: Read timed out.
```

### Screenshot

![image](https://user-images.githubusercontent.com/6815911/66565692-fe725300-eb5a-11e9-8e89-9a92a294ab74.png)

### Platform
 - [`fusionauth/fusionauth-app:1.10.1`](https://hub.docker.com/layers/fusionauth/fusionauth-app/1.10.1/images/sha256-d911eece9d296a60539ca605db976fcb4dbe15e0d513523e7630d1876fc9396c) docker image

### Additional context
To avoid this error temporarly we've increased the memory to 6GB:

![image (1)](https://user-images.githubusercontent.com/6815911/66565867-6c1e7f00-eb5b-11e9-807e-10702f928296.png)

But it's only a matter of time untils this error is thrown again.

**Comment by robotdan** [comment]
Thanks for reporting @johnmaia. A NPE should not really ever get thrown here, so perhaps we are hitting some sort of memory issue. 

This code path is not optimized for this number of tenants, so we'll need to review it further.

**Comment by robotdan** [comment]
I identified the root cause of this issue, while this code path may still not be as performant as it could be with 4k tenants, the exception is not expected. 

This will be fixed in the upcoming patch release.

**Comment by robotdan** [comment]
Available in 1.11.0

**Comment by johnmaia** [comment]
@robotdan thanks for the fast fix. Could you please release a new version of @fusionauth/node-client on npm?

**Comment by robotdan** [comment]
You're welcome @johnmaia the updated Node client should be available now. 
https://www.npmjs.com/package/@fusionauth/node-client

Side note, the Passwordless w/out our Email is also available in 1.11.0. 
https://fusionauth.io/docs/v1/tech/apis/passwordless
